### PR TITLE
Highlight all timers of the same id simultaneously

### PR DESCRIPTION
**Pull request by akopich**
Currently this works only for timers representing hooked function
events. The main obstacle is the fact that OrbitApp
and DataManager  only store id of highlighted function,
not of highlighted scope.

The timer borders are also properly drawn if histogram
selection is active.

Also, Box from Geometry.h is renamed to Tetragon, as
no Box invariants are checked. The method
Batcher::AddShadedTrapezium now takes Tetragon, not
four vectors.

Tests: Unit, Compile, Manual
Bug: http://b/227595665

### CODE REVIEW

**Comment by akopich** [comment]
http://screenshot/AsLpdPVwET4R6tH

File: src/OrbitGl/Batcher.cpp
Comment (beckerhe) [inline src/OrbitGl/Batcher.cpp:58]: The functions should probably also be renamed

Comment (beckerhe) [reply src/OrbitGl/Batcher.cpp:58]: And @karupayun should have a look how this clashes with his current refactorings

File: src/OrbitGl/OpenGlBatcher.h
Comment (beckerhe) [inline src/OrbitGl/OpenGlBatcher.h:94]: Maybe also rename this function

File: src/OrbitGl/TimerTrack.cpp
Comment (beckerhe) [inline src/OrbitGl/TimerTrack.cpp:427]: `const Vec3&`?

**Review by beckerhe — APPROVED** [review]
Otherwise LGTM. But @karupayun should also approve due to the Batcher changes.

File: src/OrbitGl/Batcher.cpp
Comment (florian-kuebler) [reply src/OrbitGl/Batcher.cpp:58]: Well actually, there even no need for the whole function anymore. AddTrapezium can handle all cases from its logic. I think AddTrapezium should be renamed to AddTetragon and add Box should be removed. OR, we keep both (Trapezium and Box), forward AddBox to AddTrapezium (to get rid of AddBoxInternal) and change AddBox to not take a Tetragon but a pos/size/z pair and calls MakeBox internally.

File: src/OrbitGl/Batcher.h
Comment (florian-kuebler) [inline src/OrbitGl/Batcher.h:143]: There should be no need for this method anymore.

File: src/OrbitGl/Batcher.cpp
Comment (florian-kuebler) [reply src/OrbitGl/Batcher.cpp:58]: Alternatively, IFF AddBoxInternal is for some reasons faster than 2xAddTriangeInternal, than rename AddBoxInternal to AddTetragonInternal and forward the AddTrapezium call to this rather than to 2xAddTriangle.

File: src/OrbitGl/ThreadTrack.cpp
Comment (florian-kuebler) [inline src/OrbitGl/ThreadTrack.cpp:214]: Thanks for adding const!

**Review by florian-kuebler — COMMENTED** [review]
In general LGTM, thanks! My main concern would be around the AddBox comment.

Also, I think this PR could (maybe should) have been split into two PRs. For me it is fine as it is, as I am done with the full review, but please keep in mind for the next time.

File: src/DataViews/LiveFunctionsDataView.cpp
Comment (akopich) [reply src/DataViews/LiveFunctionsDataView.cpp:148]: Good point. The only piece of code that relies on `selected_function_id_` is `LiveFunctionsDataView::GetRowFromFunctionId` method, which properly handles ids outside of the range of ids assigned to hooked functions. So, I'll rather remove the check.

File: src/OrbitGl/Batcher.cpp
Comment (akopich) [reply src/OrbitGl/Batcher.cpp:58]: I am not sure about renaming it. The method might expect the Tetragon to be a box semantically.

Comment (beckerhe) [reply src/OrbitGl/Batcher.cpp:58]: > I am not sure about renaming it. The method might expect the Tetragon to be a box semantically.

Then we should enforce that with the type system - not with a name.

Comment (akopich) [reply src/OrbitGl/Batcher.cpp:58]: > enforce that with the type system
But if we do, the newly created class (e.g. Box or Rectangle) will have to check whether its vertices actually form a rectangle (the box we used to have didn't do anything like that).

Comment (beckerhe) [reply src/OrbitGl/Batcher.cpp]: There are multiple options:

1. `Box` or `Rectangle` could store `pos`, `size`, and `z`. This would enforce it being a rectangle.
2. `Box` or `Rectangle` could be a class instead of a `struct` and we enforce it being a ractangle by class-invariants.

File: src/OrbitGl/TimerTrack.cpp
Comment (karupayun) [inline src/OrbitGl/TimerTrack.cpp:429]: Move this method to the class where Vec3 is defined. I also needed once.

File: src/OrbitGl/TimerTrack.cpp
Comment (karupayun) [inline src/OrbitGl/TimerTrack.cpp:438]: As I said above, this method is a clear candidate to be in the new PrimitiveAssembler (and old Batcher) class.

File: src/OrbitGl/TimerTrack.cpp
Comment (karupayun) [inline src/OrbitGl/TimerTrack.cpp:431]: If you modify AddTetragonBorder to get UserData instead of TimerInfo, this method is not needed and AddTetragonBorder could just add 4 lines (in Primitive Assembler) similar to how currently AddTrapezium works.

File: src/OrbitGl/TimerTrack.cpp
Comment (karupayun) [reply src/OrbitGl/TimerTrack.cpp:110]: Actually, isn't it simpler to include both the tetragon and also the 'z' in AddTetragonBorder method and erase this method?

**Review by karupayun — COMMENTED** [review]
Hey Valeriy, I left a few comments. Sorry for being slow at reviewing :(

**Comment by karupayun** [comment]
I also have a comment about visual elements. I remember that I suggested to not draw border to small boxes (because I could intersect with anothers showing almost everything a border), but now that I'm seeing in the screen, it's pretty confusing also that small boxes doesn't have a border. Have you ever asked Carsten about this issue? I maybe would propose to always draw the borders (lines included) but below the boxes themselves, but I would like to hear what he thinks.

File: src/OrbitGl/TimerTrack.cpp
Comment (akopich) [reply src/OrbitGl/TimerTrack.cpp:110]: Should I rather call it `GetTetragonWithNewZ` ?
I don't like to pass a Tetragon and a `float z` together anywhere. As it would effectively mean to pass two z-values and this would be a source of confusion.

**Comment by akopich** [comment]
@karupayun 
The smallest timers don't have border on the screenshot, because the histogram selection doesn't start with the minimal duration. 
But we probably do need an input from Carsten.

**Comment by karupayun** [comment]
> @karupayun The smallest timers don't have border on the screenshot, because the histogram selection doesn't start with the minimal duration. But we probably do need an input from Carsten.

Ahhh, ok!

File: src/OrbitGl/Batcher.cpp
Comment (akopich) [reply src/OrbitGl/Batcher.cpp]: We've agreed with Pablo to get rid of AddTrapezium method in a subsequent PR. AddBox seems to be capable of drawing any tetragon. TODO added.

File: src/OrbitGl/TimerTrack.cpp
Comment (karupayun) [reply src/OrbitGl/TimerTrack.cpp:110]: I have an opposite opinion. We are assuming every shape have the same z in the 4 vertices that can be not true, I would rather refactor methods to use Vec2 and add the z as a parameter instead in all the methods. This is a step backwards.

Comment (akopich) [reply src/OrbitGl/TimerTrack.cpp:110]: Agree, I never liked Vec3 to be honest. Should we at least make Tetragon store four Vec2 instances and `float z`?

**Review by florian-kuebler — APPROVED** [review]
LGTM, thanks. However, please wait until @karupayun's approval.

File: src/OrbitGl/App.cpp
Comment (akopich) [reply src/OrbitGl/App.cpp:2474]: Several PRs ago we've agreed to use IsDevMode for the whole "manual grouping" feature.

File: src/OrbitGl/Geometry.h
Comment (karupayun) [reply src/OrbitGl/Geometry.h]: Are we really requiring them to be in clockwise order? Why not asking for Vec3[4] vertices?

File: src/OrbitGl/CoreMath.h
Comment (karupayun) [inline src/OrbitGl/CoreMath.h:24]: Thanks!

**Review by karupayun — APPROVED** [review]
Thanks for addressing all my comments. I'm only not happy with Tetragon not asking for an array of 4 vertices. Otherwise LGTM!

File: src/OrbitGl/Geometry.h
Comment (akopich) [reply src/OrbitGl/Geometry.h]: Either clock-wise, or counter clock-wise. AddTetragonBorder uses this.

Comment (karupayun) [reply src/OrbitGl/Geometry.h]: I mean, for me is really obvious that if vertices are not ordered you will have a non-convex polygon, but in that case call ordered_vertices to the parameter. 

Still strongy about Array[4] vertices

Comment (akopich) [reply src/OrbitGl/Geometry.h]: btw, why isn't std::array used here?

Comment (karupayun) [reply src/OrbitGl/Geometry.h]: You can use either. I would just write vertices.

Comment (karupayun) [reply src/OrbitGl/Geometry.h]: I think we are using std::array in some methods in PrimitiveAssembler, not sure why it's not used here

**Pull request merged**
